src/serialization/ascii_char.rs

```diff
@@ -0,0 +1,75 @@
+use std::fmt;
+
+use serde::de::{Error, Unexpected, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use ascii_char::AsciiChar;
+
+impl Serialize for AsciiChar {
+    #[inline]
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        serializer.serialize_char(self.as_char())
+    }
+}
+
+struct AsciiCharVisitor;
+
+impl<'de> Visitor<'de> for AsciiCharVisitor {
+    type Value = AsciiChar;
+
+    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("an ascii character")
+    }
+
+    #[inline]
+    fn visit_char<E: Error>(self, v: char) -> Result<Self::Value, E> {
+        AsciiChar::from(v).map_err(|_| Error::invalid_value(Unexpected::Char(v), &self))
+    }
+
+    #[inline]
+    fn visit_str<E: Error>(self, v: &str) -> Result<Self::Value, E> {
+        let mut iter = v.chars();
+        match (iter.next(), iter.next()) {
```

Comment:
Wouldn't a simple check `v.len() == 1` suffice?

```
if v.len() == 1 {
    self.visit_char(v[0])
} else {
    Err(Error::invalid_value(Unexpected::Str(v), &self))
}
```
This seems simpler to me, although there exists an invariant between the first two lines.

Reply:
Although, we can't get a `char` by indexing a `&str` (i.e. no `v[0]`); so I'll do this instead...

```
        if v.len() == 1 {
            let c = v.chars().next().unwrap();
            self.visit_char(c)
        } else {
            Err(Error::invalid_value(Unexpected::Str(v), &self))
        }
```

Reply:
Unless there's a more elegant way you can think of to get that `char` from the `&str`.

Reply:
Ah, you're right.. I should have tested that at the playground.

Interestingly the `char::from_str` function looks just like your code. I'd say, just keep it your way.